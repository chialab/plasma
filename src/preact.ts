import { mkdir, writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import { capitalize, filterPublicMemebers } from './utils';
import type { Entry } from './walker';

export interface PreactTransformOptions {
    /**
     * The entrypoint to the package.
     */
    entrypoint: string;

    /**
     * The output directory to write the converted components to.
     */
    outdir: string;
}

export function generatePreactComponent(entry: Entry, options: PreactTransformOptions) {
    const { definition, declaration } = entry;

    const props = filterPublicMemebers(declaration).map((member) => member.name);
    const eventProps =
        declaration.events?.reduce(
            (acc, event) => ({
                ...acc,
                [`on${capitalize(event.name)}`]: event.name,
            }),
            {} as Record<string, string>
        ) ?? {};

    return `import { h } from 'preact';
import { useRef, useEffect } from 'preact/hooks';
import '${options.entrypoint}';

const properties = ${JSON.stringify(props)};
const events = ${JSON.stringify(eventProps)};

export const ${declaration.name} = ({ children, ...props }) => {
    const ref = useRef();
    const {
        ${props.join(',\n        ')},
        ${Object.keys(eventProps).join(',\n        ')},
        ...restProps
    } = props;

    useEffect(() => {
        const listeners = [];

        for (const prop in props) {
            if (prop in events) {
                ref.current.addEventListener(events[prop], props[prop]);
                listeners.push(() => ref.current.removeEventListener(events[prop], props[prop]));
            } else if (properties.includes(prop)) {
                ref.current[prop] = props[prop];
            }
        }

        return () => {
            listeners.forEach((listener) => listener());
        };
    });

    return h('${definition.extend ?? definition.name}', {
        ref,${definition.extend ? `\n        is: '${definition.name}',` : ''}
        ...restProps,
    }, children);
};`;
}

export function generatePreactTypings(entry: Entry, options: PreactTransformOptions) {
    const { declaration } = entry;

    const imports = `import { FunctionComponent, JSX } from 'preact';
import { ${declaration.name} as Base${declaration.name} } from '${options.entrypoint}';
`;

    const propertiesTypings = filterPublicMemebers(declaration).map(
        (member) => `${member.name}?: Base${declaration.name}['${member.name}'];`
    );

    const eventTypings =
        declaration.events?.map(
            (event) => `on${capitalize(event.name)}?: (this: Base${declaration.name}, event: CustomEvent) => boolean;`
        ) ?? [];

    return `${imports}
export const ${declaration.name}: FunctionComponent<JSX.HTMLAttributes<Base${declaration.name}> & {
    ${propertiesTypings.join('\n    ')}
    ${eventTypings.join('\n    ')}
    [key: \`data-\${string}\`]: any;
}>;`;
}

export async function transformPreact(entry: Entry, options: PreactTransformOptions) {
    const { declaration } = entry;
    const outFile = join(options.outdir, `${declaration.name}.js`);
    const declFile = join(options.outdir, `${declaration.name}.d.ts`);

    await mkdir(options.outdir, {
        recursive: true,
    });
    await Promise.all([
        writeFile(outFile, generatePreactComponent(entry, options)),
        writeFile(declFile, generatePreactTypings(entry, options)),
    ]);

    return outFile;
}
