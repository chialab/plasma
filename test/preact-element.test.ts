import { h, render } from 'preact';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { TestElement } from './src/preact/TestElement';

let host: HTMLElement;

const awaitReactRender = () => new Promise((resolve) => setTimeout(resolve, 100));

describe('Element', () => {
    beforeEach(() => {
        host = document.createElement('div');
        document.body.appendChild(host);
    });

    afterEach(() => {
        host.remove();
    });

    test('mount component', async () => {
        const onStringChange = vi.fn();
        const onClick = vi.fn((event) => event.preventDefault());
        render(
            h(TestElement, {
                'stringProp': 'test',
                'booleanProp': true,
                'numericProp': 1,
                'objectProp': { test: true },
                'data-attr': 'test',
            }),
            host
        );
        await awaitReactRender();
        const node = host.querySelector('test-element') as HTMLElement;
        expect(host.innerHTML).toMatchSnapshot();
        expect(onStringChange).not.toHaveBeenCalled();
        expect(onClick).not.toHaveBeenCalled();
        render(
            h(TestElement, {
                'onStringchange': onStringChange,
                onClick,
                'stringProp': 'changed',
                'booleanProp': true,
                'numericProp': 1,
                'objectProp': { test: true },
                'data-attr': 'test',
            }),
            host
        );
        await awaitReactRender();
        expect(host.innerHTML).toMatchSnapshot();
        expect(onStringChange).toHaveBeenCalled();
        node.click();
        expect(onClick).toHaveBeenCalled();
    });

    test('named slot', async () => {
        render(h(TestElement, {}, ['Hello', h('i', { slot: 'icon', key: 'icon' }, 'icon')]), host);
        expect(host.innerHTML).toMatchSnapshot();
    });
});
