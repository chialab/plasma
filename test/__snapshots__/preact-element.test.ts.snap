// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Element > mount component 1`] = `"<test-element data-attr="test" :scope="test-element" :defined="" stringprop="test" booleanprop="" numericprop="1"><span class="text">test</span><span class="boolean">true</span><span class="number">1</span><span class="object">{"test":true}</span></test-element>"`;

exports[`Element > mount component 2`] = `"<test-element data-attr="test" :scope="test-element" :defined="" stringprop="changed" booleanprop="" numericprop="1"><span class="text">changed</span><span class="boolean">true</span><span class="number">1</span><span class="object">{"test":true}</span></test-element>"`;

exports[`Element > named slot 1`] = `"<test-element :scope="test-element" :defined="">Hello<span class="text"></span><span class="number"></span><span class="object"></span><i slot="icon">icon</i></test-element>"`;
